primitives/src/merkle_tree/mod.rs

```diff
@@ -196,11 +198,8 @@ pub trait MerkleTreeScheme: Sized {
     /// Tree arity
     const ARITY: usize;
 
-    /// Construct a new merkle tree with given height from a data slice
-    fn from_elems(
-        height: usize,
-        elems: impl IntoIterator<Item = impl Borrow<Self::Element>>,
-    ) -> Result<Self, PrimitivesError>;
+    /// Initialize an empty Merkle tree
+    fn new(height: usize) -> Self;
 
```

Comment:
## Open to discussion/debate

A big consequence of removing `from_elems` is that we can no longer construct a struct that _only_ implements `MerkleTreeScheme` trait, which makes this trait useless. 
Sometimes, we just want a fixed-non-appendable MT. 

I understand that removing this originates from the intention to remove `Index: From<u64>` trait bound which is not pretty, but compared to the non-implementable trait, I prefer to swallow this pill.

Reply:
I prefer the new design as it looks cleaner.
Also we could rename the `AppendableMerkleTreeScheme` to `VecMerkleTreeScheme` or something to avoid confusion

Reply:
ask for more opinions here @ggutoski @jbearer @tbro 

(i have a slight preference to keeping `MTScheme::from_elems()`)

Reply:
I don't know why `from_elems` requires `From<u64>` trait bound on `Index`. Even after moving `from_elems` to `AppendableMerkleTreeScheme`, we still have a mysterious `From<u62>` trait bound. Is there no way to remove this requirement?

# On constructor methods in traits

I have not yet seen a "correct" universal rule on whether/how to put constructor methods in traits.

Example: I had a similar problem with `VidScheme` trait. Different VID schemes require different things for construction. (Example: ADVZ requires polynomial degree & KZG params. Other VID schemes might not need these things.) I did not put a constructor in `VidScheme`.

It's still possible to construct a bare `MerkleTreeScheme`, but you need to isolate construction. Example: have a tiny module `M` that exposes a merkle tree "factory" function with return type `impl MerkleTreeScheme`. Then `M` is the only place that "knows" the underlying implementation of `MerkleTreeScheme`; everything else talks to it via this trait, maintaining modularity/abstraction. It seems I'll need to do something like this for `VidScheme`.

For merkle trees, we're starting to see demand for different constructors. eg. `from_elems`, whether to require a `height` arg, etc. It's hard to accommodate all these variations in a universal constructor method.

Reply:
Another option: `from_elems` could take an `Iterator<Item = (Self::Index, impl Borrow<Self::Element>)` and force the caller to make `Index`s for you. This does not require `From<u64>`. But it is less ergonomic for the caller. Most of the time I expect caller needs to use `.enumerate()`, which is not terribly ugly.

Reply:
Oh just like the `from_kv_set` in `UniversalMerkleTreeScheme`? That's actually a good idea.`

Reply:
> Another option: `from_elems` could take an `Iterator<Item = (Self::Index, impl Borrow<Self::Element>)` and force the caller to make `Index`s for you. This does not require `From<u64>`. But it is less ergonomic for the caller. Most of the time I expect caller needs to use `.enumerate()`, which is not terribly ugly.

This is actually a good idea. Just like `from_kv_set` from `UniversalMerkleTreeScheme`.

There's a little problem, it's not easy to implement an efficient tree builder like [the one here](https://github.com/EspressoSystems/jellyfish/blob/34059730ee692a35963732b7b9f909c9546dd5c0/primitives/src/merkle_tree/internal.rs#L178) because you can't assume order on the provided tuples which is generic over `Index`.

Reply:
Can you elaborate? I don't see where an ordering is needed. Seems that the existing implementation does not use ordering either. It merely uses `From<u64>` as a way to make a new `Index`. But you have no ordering guarantee on these indices. (eg. the `From<u64>` impl could be a hash output or string or anything.)

Reply:
Maybe you need to add a uniqueness check for each `Index` from the callers iterator...?

Reply:
Ordering is implicitly needed in that implementation. As it's just repeatedly aggregate `Arity` nodes to form an internal nodes as their parents. 

Also, what if user provide `[(0, sth), (1, sth), (3, sth)]` as the input?

Reply:
As discussed offline:
- appendable merkle tree really needs `Index` to be an integer type like `usize`, so we should probably enforce `usize`
- it should be possible to maintain `from_elems` in `MerkleTreeScheme` and remove `From<u64>`. Then modify `AppendableMerkleTree` to enforce `Index = usize`.
- We might need to use `Iterator<Item = (Self::Index, impl Borrow<Self::Element>)` as described [above](https://github.com/EspressoSystems/jellyfish/pull/475#discussion_r1473324565) to implement `from_elems` for tree types that allow generic `Index`.

It would be nice to get these changes into the present PR. But if it's too much extra work then we could push it to a later PR.

Reply:
offline discussion during today's crypto meeting:

remove constructors from traits, functions that generic over traits should be receiving a struct and only use traits for behaviors, not for construction. (by Jeb, and agreed by everyone)

Reply:
Already did. Check the newest commit